src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.cpp

```diff
@@ -51,43 +65,497 @@ void attn_acc_value(float* out, float weight, T* v, size_t S) {
+
+        auto v0_out = mm512_uni_loadu_ps(out + i);
+        auto v1_out = mm512_uni_loadu_ps(out + i + vec_len_f32_avx512);
+        auto v2_out = mm512_uni_loadu_ps(out + i + vec_len_f32_avx512 * 2);
+        auto v3_out = mm512_uni_loadu_ps(out + i + vec_len_f32_avx512 * 3);
+
+        auto v0_256 = _mm512_cvtepu8_epi32(v0_128);
+        auto v1_256 = _mm512_cvtepu8_epi32(v1_128);
+        auto v2_256 = _mm512_cvtepu8_epi32(v2_128);
+        auto v3_256 = _mm512_cvtepu8_epi32(v3_128);
+
+        auto v0_value = _mm512_cvtepi32_ps(v0_256);
+        auto v1_value = _mm512_cvtepi32_ps(v1_256);
+        auto v2_value = _mm512_cvtepi32_ps(v2_256);
+        auto v3_value = _mm512_cvtepi32_ps(v3_256);
+
+        v0_value = _mm512_sub_ps(v0_value, v_zp);
+        v1_value = _mm512_sub_ps(v1_value, v_zp);
+        v2_value = _mm512_sub_ps(v2_value, v_zp);
+        v3_value = _mm512_sub_ps(v3_value, v_zp);
+
+        v0_out = _mm512_fmadd_ps(attn_w_vec_fp32, v0_value, v0_out);
+        v1_out = _mm512_fmadd_ps(attn_w_vec_fp32, v1_value, v1_out);
+        v2_out = _mm512_fmadd_ps(attn_w_vec_fp32, v2_value, v2_out);
+        v3_out = _mm512_fmadd_ps(attn_w_vec_fp32, v3_value, v3_out);
+
+        mm512_uni_storeu_ps(out + i + vec_len_f32_avx512 * 0, v0_out);
+        mm512_uni_storeu_ps(out + i + vec_len_f32_avx512 * 1, v1_out);
+        mm512_uni_storeu_ps(out + i + vec_len_f32_avx512 * 2, v2_out);
+        mm512_uni_storeu_ps(out + i + vec_len_f32_avx512 * 3, v3_out);
+    }
+    if (i + 2 * vec_len_f32_avx512 <= S) {
+        auto v0_128 = _mm_loadu_si128(reinterpret_cast<__m128i*>(v + i));
+        auto v1_128 = _mm_loadu_si128(reinterpret_cast<__m128i*>(v + i + vec_len_f32_avx512));
+
+        auto v0_out = mm512_uni_loadu_ps(out + i);
+        auto v1_out = mm512_uni_loadu_ps(out + i + vec_len_f32_avx512);
+
+        auto v0_256 = _mm512_cvtepu8_epi32(v0_128);
+        auto v1_256 = _mm512_cvtepu8_epi32(v1_128);
+
+        auto v0_value = _mm512_cvtepi32_ps(v0_256);
+        auto v1_value = _mm512_cvtepi32_ps(v1_256);
+
+        v0_value = _mm512_sub_ps(v0_value, v_zp);
+        v1_value = _mm512_sub_ps(v1_value, v_zp);
+
+        v0_out = _mm512_fmadd_ps(attn_w_vec_fp32, v0_value, v0_out);
+        v1_out = _mm512_fmadd_ps(attn_w_vec_fp32, v1_value, v1_out);
+
+        mm512_uni_storeu_ps(out + i + vec_len_f32_avx512 * 0, v0_out);
+        mm512_uni_storeu_ps(out + i + vec_len_f32_avx512 * 1, v1_out);
+        i += 2 * vec_len_f32_avx512;
+    }
+    if (i + vec_len_f32_avx512 <= S) {
+        auto v0_128 = _mm_loadu_si128(reinterpret_cast<__m128i*>(v + i));
+        auto v0_out = mm512_uni_loadu_ps(out + i);
+        auto v0_256 = _mm512_cvtepu8_epi32(v0_128);
+        auto v0_value = _mm512_cvtepi32_ps(v0_256);
+        v0_value = _mm512_sub_ps(v0_value, v_zp);
+        v0_out = _mm512_fmadd_ps(attn_w_vec_fp32, v0_value, v0_out);
+        mm512_uni_storeu_ps(out + i + vec_len_f32_avx512 * 0, v0_out);
+        i += vec_len_f32_avx512;
```

Comment:
2 questions:
1. Does all this manual unrolling on factors 4 and 2 give any noticable perf improvemen?
2. As attempt to minimize code duplication can we try to extend marked code lines into inlide function and reuse it by multiple calls inside unrolled loops?

Reply:
1, It should be. Generally, if there is data dependency manually unrolling will break the limit and throughput will be increased, `dot_product` is this category. When the kernel is very small, unrolling may decrease the branch cost, `dot_product` and `attn_acc_value` are this class, especially when the memory access time may close to the computation time, so avoid any waste should be important.
2, Because the kernels are small and inline is very important, but I found the compiler may not do the inline as expected, such as `dot_product` from my test. May I leave this as a legacy issue and try to solve it in the next iteration optimization about it?

Reply:
> May I leave this as a legacy issue and try to solve it in the next iteration optimization?

For sure. That is not critical, Just an intention to investigate possible ways to optimize loop unrolling procedure for our intrinsics kernels. Should bring good value in case number of intrinsics kernels will grow.